src/ert/scheduler/job.py

```diff
@@ -63,41 +66,58 @@ def driver(self) -> Driver:
         return self._scheduler.driver
 
     async def __call__(
-        self, start: asyncio.Event, sem: asyncio.BoundedSemaphore
+        self, start: asyncio.Event, sem: asyncio.BoundedSemaphore, max_submit: int = 2
     ) -> None:
         await start.wait()
-        await sem.acquire()
-
-        try:
-            await self._send(State.SUBMITTING)
-            await self.driver.submit(
-                self.real.iens, self.real.job_script, cwd=self.real.run_arg.runpath
-            )
-
-            await self._send(State.STARTING)
-            await self.started.wait()
-
-            await self._send(State.RUNNING)
-            while not self.returncode.done():
-                await asyncio.sleep(0.01)
-            returncode = await self.returncode
-            if (
-                returncode == 0
-                and forward_model_ok(self.real.run_arg).status
-                == LoadStatus.LOAD_SUCCESSFUL
-            ):
-                await self._send(State.COMPLETED)
-            else:
-                await self._send(State.FAILED)
-
-        except asyncio.CancelledError:
-            await self._send(State.ABORTING)
-            await self.driver.kill(self.iens)
-
-            await self.aborted.wait()
-            await self._send(State.ABORTED)
-        finally:
-            sem.release()
+        retries = 0
+        retry: bool = True
+        while retry:
+            retry = False
+            await sem.acquire()
+            try:
+                await self._send(State.SUBMITTING)
+                await self.driver.submit(
+                    self.real.iens, self.real.job_script, cwd=self.real.run_arg.runpath
+                )
+
+                await self._send(State.STARTING)
+                await self.started.wait()
+
+                await self._send(State.RUNNING)
+                while not self.returncode.done():
+                    await asyncio.sleep(0.01)
+                returncode = await self.returncode
+                # we need to make sure that the task has finished too
+                await self.driver.wait(self.real.iens)
```

Comment:
Is it optimal to use the name `wait()` for this, as it looks like the same as the lines with `self.started.wait()`, while the latter is waiting for an asyncio.Event.

Reply:
Maybe a name sufficiently clear to avoid the comment.

Reply:
> Why can't the LocalDriver itself do the right thing when the job is ended / restarted?

Since we want to have it non-blocking and thus we need to await out of the scope. The thing that Subprocess has ended does not guarantee that the task has ended too, ie. got failures on tests.

Reply:
The fact that it's a subprocess is implementation detail. `await self.returncode` already means "wait for the job to end". Do we need another `wait` function to do this?

Reply:
Correct, but as you said it's the job that has finished not the task. You can try to remove it and the test with retries will fail. I guess this `await task` does more to that async context.